### PR TITLE
[Lang] [bug] Add ti.pass_value() annotation for matrix operations to prevent duplicate evaluation

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #1424 https://github.com/taichi-dev/taichi/issues/1416#issuecomment-654379663

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: tests/python/test_ssa.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+
+@ti.all_archs
+def test_func_random_argument_dup_eval():
+    @ti.func
+    def func(a):
+        return ti.Vector([ti.cos(a), ti.sin(a)])
+
+    @ti.kernel
+    def kern(t: ti.f32) -> ti.f32:
```
Comment (xumingkuan) [inline tests/python/test_ssa.py:70]: Looks like `t` is unused?

**Comment by archibate** [comment]
> It seems that function arguments are passed by value by default?

Recall that classfunc pass `self` as `ti.template()` by default. (by @k-ye and @KLozes)

File: python/taichi/lang/kernel_arguments.py [hunk 2]
```diff
@@ -32,6 +32,20 @@ def extract(self, x):
...
+        self.dim = dim
+
+    def extract(self, x):
+        import taichi as ti
+        if isinstance(x, ti.Expr):
+            return ti.expr_init(x)
+        return x
+
+
+pass_value = PassValue
```
[the remark this answers is not in the document]
Comment (archibate) [reply python/taichi/lang/kernel_arguments.py:46]: Yes, let's do the refactor iapr. Feel free to open an issue to track this.

**Comment by archibate** [comment]
> LGTMig.
> 
> I thought duplicate evaluation prevention is for #1329 but it looks like that issue is already solved?

No, actually `ti.random()` will be duplicated in vector initializer before this PR:
```
root@archlinux ~/taichi (git)-[master] # pytest ssa.py 
==================================== test session starts =====================================
platform linux -- Python 3.8.3, pytest-5.4.3, py-1.8.2, pluggy-0.13.1
rootdir: /root/taichi
plugins: forked-1.1.3, rerunfailures-9.0, xdist-1.32.0
collected 4 items                                                                            

ssa.py .F..                                                                            [100%]

========================================== FAILURES ==========================================
________________________________ test_random_vector_dup_eval _________________________________

    @ti.all_archs
    def test_random_vector_dup_eval():
        a = ti.Vector(2, ti.f32, ())
    
        @ti.kernel
        def func():
            a[None] = ti.Vector([ti.random(), 1]).normalized()
    
        for i in range(4):
            func()
>           assert a[None].value.norm_sqr() == approx(1)
E           assert 0.5803180595854402 == 1 ± 1.0e-06
E            +  where 0.5803180595854402 = <bound method Matrix.norm_sqr of [0.08082835 0.75748587]>()
E            +    where <bound method Matrix.norm_sqr of [0.08082835 0.75748587]> = [0.08082835 0.75748587].norm_sqr
E            +      where [0.08082835 0.75748587] = <taichi.lang.matrix.Matrix.Proxy object at 0x7f79447ef0a0>.value
E            +  and   1 ± 1.0e-06 = approx(1)

ssa.py:46: AssertionError
------------------------------------ Captured stdout call ------------------------------------
Running test on arch=Arch.x64
[Taichi] Starting on arch=x64
================================== short test summary info ===================================
FAILED ssa.py::test_random_vector_dup_eval - assert 0.5803180595854402 == 1 ± 1.0e-06
================================ 1 failed, 3 passed in 4.05s =================================
1 root@archlinux ~/taichi (git)-[master] # gg norm_sqr-use-self-dot-self                   :(
Switched to branch 'norm_sqr-use-self-dot-self'
Your branch is ahead of 'upstream/norm_sqr-use-self-dot-self' by 1 commit.
  (use "git push" to publish your local commits)
root@archlinux ~/taichi (git)-[norm_sqr-use-self-dot-self] # pytest ssa.py
==================================== test session starts =====================================
platform linux -- Python 3.8.3, pytest-5.4.3, py-1.8.2, pluggy-0.13.1
rootdir: /root/taichi
plugins: forked-1.1.3, rerunfailures-9.0, xdist-1.32.0
collected 4 items                                                                            

ssa.py ....                                                                            [100%]

===================================== 4 passed in 3.58s ======================================
root@archlinux ~/taichi (git)-[norm_sqr-use-self-dot-self] # 
```

**Comment by archibate** [comment]
```py
import taichi as ti
ti.init(print_preprocessed=True)

@ti.kernel
def func():
    v = ti.Vector([2, 3])
    v.norm_sqr()

func()
```
Before this PR:
```
Before preprocessing:
def norm_sqr(self):
  return self.dot(self)

After preprocessing:
def norm_sqr(self):
  import taichi as ti
  return self.dot(self)
```
After this PR:
```
Before preprocessing:
def norm_sqr(self: impl.pass_value()):
  return self.dot(self)

After preprocessing:
def norm_sqr(self_by_value__: impl.pass_value()):
  import taichi as ti
  self = ti.expr_init_func(self_by_value__)
  return self.dot(self)Before preprocessing:
def norm_sqr(self: impl.pass_value()):
  return self.dot(self)

After preprocessing:
def norm_sqr(self_by_value__: impl.pass_value()):
  import taichi as ti
  self = ti.expr_init_func(self_by_value__)
  return self.dot(self)
```

**Comment by xumingkuan** [comment]
> After this PR:
> 
> ```
> Before preprocessing:
> def norm_sqr(self: impl.pass_value()):
>   return self.dot(self)
> 
> After preprocessing:
> def norm_sqr(self_by_value__: impl.pass_value()):
>   import taichi as ti
>   self = ti.expr_init_func(self_by_value__)
>   return self.dot(self)Before preprocessing:
> def norm_sqr(self: impl.pass_value()):
>   return self.dot(self)
> 
> After preprocessing:
> def norm_sqr(self_by_value__: impl.pass_value()):
>   import taichi as ti
>   self = ti.expr_init_func(self_by_value__)
>   return self.dot(self)
> ```

@archibate Did you paste it twice or it printed twice?

**Comment by archibate** [comment]
paste twice, sorry!

**Comment by archibate** [comment]
Hello? May I merge this now? I'm afraid it will soon cause conflict.

**Comment by yuanming-hu** [comment]
I'm slightly confused by "SSA" when we are talking about the frontend. Taichi's frontend language is not SSA, so I'm not sure what that refers to ...

**Comment by archibate** [comment]
> I'm slightly confused by "SSA" when we are talking about the frontend. Taichi's frontend language is not SSA, so I'm not sure what that refers to ...

I thought it's a sum up of all kind of problems due to the expr system including:
- self-assign order
- duplicate evaluation
- ...

**Pull request closed without merging**